core/spec/services/spree/cart/change_currency_spec.rb

```diff
@@ -2,26 +2,89 @@
 
 module Spree
   describe Cart::ChangeCurrency do
-    subject { described_class.call(order: order, new_currency: new_currency) }
+    context 'when shipment is already choosen' do
+      let(:country) { create(:country) }
+      let(:store) { create(:store, checkout_zone: zone, supported_currencies: 'USD,EUR') }
+      let(:order2) { create(:order_with_line_items, store: store, ship_address: create(:address, country: country)) }
 
-    let(:order) { create(:order_with_line_items, store: store, currency: 'USD') }
-    let(:store) { create(:store, supported_currencies: 'USD,EUR,GBP') }
+      let(:zone) { create(:zone_with_country) }
 
-    context 'when switching to a supported currency' do
-      let(:new_currency) { 'EUR' }
+      before do
+        zone.countries << country
+      end
 
-      context 'when product has a price in given currency' do
-        let!(:price) { create(:price, currency: 'EUR', variant: order.line_items.first.variant)}
 
-        it 'changes order and line items currency' do
-          expect(subject).to be_success
-          order.reload
-          expect(order.currency).to eq('EUR')
-          expect(order.line_items.first.currency).to eq('EUR')
+      let!(:price) { create(:price, currency: 'EUR', variant: order2.line_items.first.variant,amount: 25.00 )}
+
+      context 'when in address state' do
+        it 'resets the checkout process when in address state' do
+          expect(Cart::ChangeCurrency.call(order: order2,new_currency: 'EUR')).to be_success
+          order2.reload
+          expect(order2.currency).to eq('EUR')
+          expect(order2.total).to eq(25.00)
+          expect(order2.shipments.to_a).to eq([])
         end
       end
 
-      context 'when product does not have a price in given currency' do
+      context 'when in payment state' do
+        it 'resets the checkout process when in payment state' do
+          Checkout::Advance.call(order: order2)
+          order2.item_total = 10
+          expect(Cart::ChangeCurrency.call(order: order2,new_currency: 'EUR')).to be_success
+          order2.reload
+          expect(order2.state).to eq('address')
+          expect(order2.currency).to eq('EUR')
+          expect(order2.total).to eq(25.00)
+          expect(order2.shipments.to_a).to eq([])
+        end
+      end
+
+      context 'if the order has shipping methods after change' do
```

Comment:
Not sure if I follow this logic, will it have shipments after the change or not?

Reply:
it will after the GetShippingRates is called(it generates a generic shipment) but the shipment will be different from the previous one  - from the customer side it would like like: get products into cart -> go to checkout -> choose shipping method -> change currency to EUR -> check if user can choose shipment method and if the shipment method is not the same as used in previous currency

Reply:
Okay, could you make the name of this context more clear? The 'if' sounds like there's something that you're not sure of inside the test. Something like `when user requests shipping methods after changing currency`

https://www.betterspecs.org/ <- here's a nice guideline on structuring and naming test cases